### DISCUSSION TITLE
github.dev: Check if file names contain illegal characters

### DISCUSSION THREAD

**Discussion by unrealapex** [community/community#16011] [Codespaces]
An issue I have encountered lately is being unable to clone a git repos which I primarily work on in the `github.dev` editor because some of the files in the repo contain characters that are illegal in file names like colon(`:`) for example. It would be nice if the web editor checked file names and warned the user if they contained any illegal characters or prevent them from creating it until the illegal characters were removed.

**Comment by dmilburn** [comment]
Thanks so much for the feedback! And you definitely describe a better user experience, I'm sorry for the inconvenience. I've added this to our internal backlog to get prioritized.

**Comment by jkeech** [comment]
@UnrealApex, different filesystems support different sets of characters in file and directory names, and Git doesn't try to restrict the set of allowed characters such that it is guaranteed to work on all filesystems. There are some paths that are valid in Git, which are therefore also supported in GitHub and `github.dev`, but which cannot be successfully checked out on all filesystems.

In particular, if you are on Windows with an NTFS file system, you cannot create a file with a colon `:` in the name. But on linux and macOS, `:` is valid in the file name. If you commit and push a repo on GitHub that includes a path with a colon in it, you'll be able to checkout the file on linux and macOS, but you won't be able to checkout that file on Windows.

The same is true for case-sensitive vs case-insensitive file systems. If you create a file named `foo` and another named `FOO` on a case-sensitive filesystem and commit them, it will be valid in Git as two distinct file paths. If you then try to checkout those files on a case-insensitive filesystem, one will overwrite the other.

### Examples
As an example, the following is supported in Git when running on a linux system:
```bash
echo test > test:file
git add test:file
git commit -m "test"
```

You can also go through the GitHub.com UI and create a file with a colon in it in the same way.
![image](https://user-images.githubusercontent.com/746020/169605263-b9b97f1f-2e32-424f-a0b8-c9c7b13db545.png)

![image](https://user-images.githubusercontent.com/746020/169605377-001a95a4-941e-47fc-8cda-b3476cd10c52.png)


Since Git allows those files to be created, I think it's the correct behavior that `github.dev` also allows those files to be created in the repo which is consistent with other Git tooling.

**Reply by unrealapex** [reply]
Thank you for your informative and detailed response! Perhaps adding a warning that certain characters are invalid in file names for certain filesystems in `github.dev`? Though I think that adding this feature would be out of scope here and would be better suited as an issue in the [Visual Studio Code repo](https://github.com/microsoft/vscode).

**Reply by jkeech** [reply]
Yeah, I think adding something more holistically to GitHub would benefit this scenario, such as a repo setting that disallows commits with characters in paths/branch names/tag names that are invalid on some popular file systems. If that repo setting were enabled, `git push` would reject commits with those paths. That would make every client, including `github.dev`, `github.com` UI, `git` on the command line, and VS Code's Git integration, all fail to push those commits and then be able to surface a friendly message back to the user saying those paths are not allowed since they might not work on all systems.

The primary benefit of doing it at the repo level is that every Git client behaves the same way so you get consistency no matter which client people use. That would prevent the bad paths from sneaking in to the repo since it's blocked at the `git push` layer. Some other Git hosting providers offer these types of repo settings that people can enable: https://docs.microsoft.com/en-us/azure/devops/repos/git/os-compatibility?view=azure-devops#path-and-file-name-restrictions.
